src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/cache_rotation.hpp

```diff
@@ -0,0 +1,234 @@
+
+            float* current_rotation_coeffts_cos_ptr = current_rotation_coeffts_ptr;
+            float* current_rotation_coeffts_sin_ptr = current_rotation_coeffts_ptr + embedding_size / 2;
+
+            for (size_t iter_idx = 0; iter_idx < num_iterations; iter_idx++,
+                        current_x_values_ptr += vec_len_in_f32_elts,
+                        current_y_values_ptr += vec_len_in_f32_elts,
+                        current_rotation_coeffts_cos_ptr += vec_len_in_f32_elts,
+                        current_rotation_coeffts_sin_ptr += vec_len_in_f32_elts) {
+#    if defined(HAVE_AVX512F)
+                rotate_kv_cache_chunk_avx512(current_x_values_ptr,
+                                             current_y_values_ptr,
+                                             current_rotation_coeffts_cos_ptr,
+                                             current_rotation_coeffts_sin_ptr,
+                                             num_processed_elements_per_iteration / 2,
+                                             is_tail);
+#    else   // HAVE_AVX2
+                rotate_kv_cache_chunk_avx2(current_x_values_ptr,
+                                           current_y_values_ptr,
+                                           current_rotation_coeffts_cos_ptr,
+                                           current_rotation_coeffts_sin_ptr,
+                                           num_processed_elements_per_iteration / 2,
+                                           is_tail);
+#    endif  // defined(HAVE_AVX512F)
+            }
+        }
+    }
+#endif      // !defined(HAVE_AVX512F) && !defined(HAVE_AVX2F)
+}
+
+template <class CT>
+inline static void rotate_kv_cache_block_ref(CT* cache_block_ptr,
+                                            float* block_rotation_coefficients_ptr,
+                                            size_t num_heads,
+                                            size_t block_size,
+                                            size_t embedding_size) {
+    for (size_t head_idx = 0; head_idx < num_heads; head_idx++) {
+        for (size_t tok_idx = 0; tok_idx < block_size; tok_idx++) {
+            size_t token_offset = embedding_size * tok_idx;
+            CT* token_embedding_data_start_in_cache =
+                cache_block_ptr + head_idx * embedding_size * block_size + embedding_size * tok_idx;
+            float* token_data_start_in_rotation_coefficients = block_rotation_coefficients_ptr + token_offset;
+            for (size_t embedding_pair_idx = 0; embedding_pair_idx < embedding_size / 2; embedding_pair_idx++) {
+                // NB: below is the llama-style rotation (x-like values are in the first half of the embedding vector,
+                // y-like values are in the second half), which is different from the original RoFormer style (x- and y-
+                // values are interleaved), but still preserves the relative positional encoding property
```

Comment:
> but still preserves the relative positional encoding property

Do you mean it correctly works even if a model uses interleaved mode?

Reply:
The invariant only holds if we consider the use case starting from model-train time, i.e. models with RoFormer-like, "interleaved" RoPE are supposed to be able to get trained up to roughly the same accuracy as the models with Llama-like "rotate_half" RoPE. At runtime the two are not interchangeable, of course.

Right now that is also a limitation of this cache rotation implementation - only the `rotate_half` version is supported, and for the proper functioning across models we should implement the rest of the RoPE variants like in https://github.com/openvinotoolkit/openvino/blob/571e98d5880a30e9d8ca25f445c343e955e79123/src/plugins/intel_cpu/src/nodes/rope.cpp#L95, with detection of the proper variant based on model metadata.